### DISCUSSION TITLE
Multiplatform libraries for Avalonia

### DISCUSSION THREAD

**Discussion by buzzware** [AvaloniaUI/Avalonia#13348] [Q&A]
1) How should multiplatform libraries be organised for Avalonia ?

a) multiple packages ie MyLibrary & MyLibrary.Android, MyLibrary.iOS etc ?
b) multitargetting ?

2) Does Avalonia support platform compiler conditionals like \_\_ANDROID\_\_ ?

**Comment by timunie** [comment]
1. I'd use several projects, one per each platform. 99 % goes into the main project
2. For platform specific code I define an interface in the main prj and implement it per platform
3. Check also the XPLAT tempalte we provide

Happy coding

**Comment by thevortexcloud** [comment]
In addition to what Tim said. I also suggest avoid using TFM specific targets if you can. EG `.net6.0-windows`. It can make compiling a pain depending on how you structure your code (as the compiler will refuse to compile the project for use in the generic target projects, even if the code can be used generically). But that can be fairly hard to avoid unless you feel like implementing all the native code interop yourself and just doing your own checks at compile/runtime. If you are targeting android/ios though, you don't really have a choice here. But If you don't mind having an executable per platform then using the platform specific TFM is a valid option.

If you are sticking with a single executable then Dependency injection can be fairly useful here too. You just switch what services you want to use on boot. For more complex things/things that must be disposed of after use you can also create factory interfaces and inject those instead.

As for multi targeting versions of .NET it can depend. If you need to ensure .NET framework compatibility you should target .NET standard 2.0 for all your library projects. Avalonia does not directly support running in a .NET standard 2.0 project so you need to reference it from a .NET Framework or .NET Core/5/6/7/8 project. But you can use the 2.0 project for most of your application logic.

You can define your own platform specific preprocesor directives fairy easily using msbuild properties. But personally I prefer using things like `OperatingSystem.IsWindows()` at runtime. It makes it far easier to catch things like build errors as you are developing that might be missed completely if you normally only compile for a single platform.

**Comment by maxkatz6** [comment]
To help make it easier, Avalonia doesn't maintain its own mobile SDK or tooling. 
This means you can both multiproject and multitarget (as far as your target platforms allow it, i.e. there is no linux TFM for example). 
And yes, you have ANDROID directive defined by .NET SDK in net7.0-android projects.
So, it's pretty standard now.
